File: windows/VpnRouter.Vpn/WireGuard/WireGuardConfigParser.cs
using System.Net;

namespace VpnRouter.Vpn.WireGuard;

public sealed class WireGuardConfigParser
{
    private const string InterfaceSection = "Interface";
    private const string PeerSection = "Peer";

    public WireGuardConfig Parse(string configText)
    {
        if (string.IsNullOrWhiteSpace(configText))
        {
            throw new WireGuardConfigException("WireGuard config is empty.");
        }

        var sections = ParseSections(configText);
        var interfaceValues = GetRequiredSection(sections, InterfaceSection);
        var peers = sections
            .Where(section => string.Equals(section.Name, PeerSection, StringComparison.OrdinalIgnoreCase))
            .Select(ParsePeer)
            .ToArray();

        if (peers.Length == 0)
        {
            throw new WireGuardConfigException("WireGuard config must contain at least one [Peer] section.");
        }

        var parsed = new WireGuardConfig(ParseInterface(interfaceValues), peers);
        Validate(parsed);
        return parsed;
    }

    public WireGuardImportResult PrepareImport(string configText)
    {
        var config = Parse(configText);
        var sanitizedConfig = Sanitize(configText);

        return new WireGuardImportResult(
            config,
            sanitizedConfig,
            config.Interface.PrivateKey,
            CreateSummary(config));
    }

    public string Sanitize(string configText)
    {
        if (string.IsNullOrWhiteSpace(configText))
        {
            return string.Empty;
        }

        var sanitizedLines = new List<string>();
        foreach (var rawLine in NormalizeLineEndings(configText).Split('\n'))
        {
            var line = rawLine.TrimEnd('\r');
            var trimmed = line.Trim();
            if (trimmed.StartsWith("PrivateKey", StringComparison.OrdinalIgnoreCase)
                && TrySplitKeyValue(trimmed, out var key, out _))
            {
                sanitizedLines.Add($"{key} = <stored securely>");
            }
            else
            {
                sanitizedLines.Add(line);
            }
        }

        return string.Join(Environment.NewLine, sanitizedLines).Trim();
    }

    public WireGuardConfigSummary CreateSummary(WireGuardConfig config)
    {
        return new WireGuardConfigSummary(
            HasInterfaceSection: true,
            HasPeerSection: config.Peers.Count > 0,
            HasPrivateKey: !string.IsNullOrWhiteSpace(config.Interface.PrivateKey),
            InterfaceAddresses: config.Interface.Addresses,
            DnsServers: config.Interface.DnsServers,
            PeerCount: config.Peers.Count,
            AllowedIps: config.Peers.SelectMany(peer => peer.AllowedIps).Distinct(StringComparer.OrdinalIgnoreCase).ToArray(),
            Endpoints: config.Peers.Select(peer => peer.Endpoint).Where(endpoint => endpoint is not null).Cast<string>().ToArray());
    }

    private static WireGuardInterfaceConfig ParseInterface(ParsedSection section)
    {
        var values = section.Values;

        return new WireGuardInterfaceConfig(
            PrivateKey: GetRequiredValue(values, "PrivateKey", InterfaceSection),
            Addresses: SplitCsv(values.GetValueOrDefault("Address")).ToArray(),
            DnsServers: SplitCsv(values.GetValueOrDefault("DNS")).ToArray(),
            ListenPort: ParseNullableInt(values.GetValueOrDefault("ListenPort"), "ListenPort"));
    }

    private static WireGuardPeerConfig ParsePeer(ParsedSection section)
    {
        var values = section.Values;

        return new WireGuardPeerConfig(
            PublicKey: GetRequiredValue(values, "PublicKey", PeerSection),
            PresharedKey: values.GetValueOrDefault("PresharedKey"),
            Endpoint: values.GetValueOrDefault("Endpoint"),
            AllowedIps: SplitCsv(values.GetValueOrDefault("AllowedIPs")).ToArray(),
            PersistentKeepalive: ParseNullableInt(values.GetValueOrDefault("PersistentKeepalive"), "PersistentKeepalive"));
    }

    private static void Validate(WireGuardConfig config)
    {
        if (config.Interface.Addresses.Count == 0)
        {
            throw new WireGuardConfigException("[Interface] must contain at least one Address.");
        }

        ValidateBase64LikeKey(config.Interface.PrivateKey, "PrivateKey");

        foreach (var address in config.Interface.Addresses)
        {
            ValidateCidr(address, "Address");
        }

        foreach (var dnsServer in config.Interface.DnsServers)
        {
            if (!IPAddress.TryParse(dnsServer, out _))
            {
                throw new WireGuardConfigException($"DNS value is not a valid IP address: {dnsServer}");
            }
        }

        foreach (var peer in config.Peers)
        {
            ValidateBase64LikeKey(peer.PublicKey, "PublicKey");

            if (!string.IsNullOrWhiteSpace(peer.PresharedKey))
            {
                ValidateBase64LikeKey(peer.PresharedKey, "PresharedKey");
            }

            if (peer.AllowedIps.Count == 0)
            {
                throw new WireGuardConfigException("[Peer] must contain at least one AllowedIPs value.");
            }

            foreach (var allowedIp in peer.AllowedIps)
            {
                ValidateCidr(allowedIp, "AllowedIPs");
            }
        }
    }

    private static IReadOnlyList<ParsedSection> ParseSections(string configText)
    {
        var sections = new List<ParsedSection>();
        ParsedSection? currentSection = null;

        var lineNumber = 0;
        foreach (var rawLine in NormalizeLineEndings(configText).Split('\n'))
        {
            lineNumber++;
            var line = StripComment(rawLine).Trim();
            if (string.IsNullOrWhiteSpace(line))
            {
                continue;
            }

            if (line.StartsWith('[') && line.EndsWith(']'))
            {
                var sectionName = line[1..^1].Trim();
                if (string.IsNullOrWhiteSpace(sectionName))
                {
                    throw new WireGuardConfigException($"Empty section name at line {lineNumber}.");
                }

                currentSection = new ParsedSection(sectionName, new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase));
                sections.Add(currentSection);
                continue;
            }

            if (currentSection is null)
            {
                throw new WireGuardConfigException($"Key/value pair appears before any section at line {lineNumber}.");
            }

            if (!TrySplitKeyValue(line, out var key, out var value))
            {
                throw new WireGuardConfigException($"Invalid key/value line at line {lineNumber}.");
            }

            currentSection.Values[key] = value;
        }

        return sections;
    }

    private static ParsedSection GetRequiredSection(IReadOnlyList<ParsedSection> sections, string name)
    {
        var section = sections.FirstOrDefault(section => string.Equals(section.Name, name, StringComparison.OrdinalIgnoreCase));
        return section ?? throw new WireGuardConfigException($"WireGuard config must contain [{name}] section.");
    }

    private static string GetRequiredValue(IReadOnlyDictionary<string, string> values, string key, string sectionName)
    {
        return values.TryGetValue(key, out var value) && !string.IsNullOrWhiteSpace(value)
            ? value
            : throw new WireGuardConfigException($"[{sectionName}] must contain {key}.");
    }

    private static bool TrySplitKeyValue(string line, out string key, out string value)
    {
        var equalsIndex = line.IndexOf('=');
        if (equalsIndex <= 0)
        {
            key = string.Empty;
            value = string.Empty;
            return false;
        }

        key = line[..equalsIndex].Trim();
        value = line[(equalsIndex + 1)..].Trim();
        return !string.IsNullOrWhiteSpace(key);
    }

    private static IEnumerable<string> SplitCsv(string? value)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return [];
        }

        return value
            .Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
            .Where(item => !string.IsNullOrWhiteSpace(item));
    }

    private static int? ParseNullableInt(string? value, string key)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return null;
        }

        return int.TryParse(value, out var parsed) && parsed is >= 0 and <= 65535
            ? parsed
            : throw new WireGuardConfigException($"{key} must be a valid port/interval integer.");
    }

    private static void ValidateBase64LikeKey(string value, string key)
    {
        if (value.Length < 32)
        {
            throw new WireGuardConfigException($"{key} is too short to be a valid WireGuard key.");
        }
    }

    private static void ValidateCidr(string value, string key)
    {
        var parts = value.Split('/', StringSplitOptions.TrimEntries);
        if (parts.Length != 2 || !IPAddress.TryParse(parts[0], out _) || !int.TryParse(parts[1], out var prefix))
        {
            throw new WireGuardConfigException($"{key} value must be CIDR notation: {value}");
        }

        var maxPrefix = parts[0].Contains(':') ? 128 : 32;
        if (prefix < 0 || prefix > maxPrefix)
        {
            throw new WireGuardConfigException($"{key} prefix is out of range: {value}");
        }
    }

    private static string StripComment(string line)
    {
        var hashIndex = line.IndexOf('#');
        return hashIndex >= 0 ? line[..hashIndex] : line;
    }

    private static string NormalizeLineEndings(string value) => value.Replace("\r\n", "\n").Replace('\r', '\n');

    private sealed record ParsedSection(string Name, Dictionary<string, string> Values);
}

public sealed record WireGuardImportResult(
    WireGuardConfig Config,
    string SanitizedConfig,
    string PrivateKeySecret,
    WireGuardConfigSummary Summary);

public sealed record WireGuardConfig(WireGuardInterfaceConfig Interface, IReadOnlyList<WireGuardPeerConfig> Peers);

public sealed record WireGuardInterfaceConfig(
    string PrivateKey,
    IReadOnlyList<string> Addresses,
    IReadOnlyList<string> DnsServers,
    int? ListenPort);

public sealed record WireGuardPeerConfig(
    string PublicKey,
    string? PresharedKey,
    string? Endpoint,
    IReadOnlyList<string> AllowedIps,
    int? PersistentKeepalive);

public sealed record WireGuardConfigSummary(
    bool HasInterfaceSection,
    bool HasPeerSection,
    bool HasPrivateKey,
    IReadOnlyList<string> InterfaceAddresses,
    IReadOnlyList<string> DnsServers,
    int PeerCount,
    IReadOnlyList<string> AllowedIps,
    IReadOnlyList<string> Endpoints);

public sealed class WireGuardConfigException(string message) : Exception(message);

File: macos/VPNRouter/VPNRouter/Shared/WireGuard/WireGuardConfigParser.swift
import Foundation
import Network

enum WireGuardConfigParser {
    private nonisolated static let interfaceSection = "Interface"
    private nonisolated static let peerSection = "Peer"

    nonisolated static func parse(_ configText: String) throws -> WireGuardConfig {
        guard !configText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw WireGuardConfigError.invalid("WireGuard 설정 파일이 비어 있습니다.")
        }

        let sections = try parseSections(configText)
        let interfaceValues = try requiredSection(sections, named: interfaceSection)
        let peers = try sections
            .filter { $0.name.caseInsensitiveCompare(peerSection) == .orderedSame }
            .map(parsePeer)

        guard !peers.isEmpty else {
            throw WireGuardConfigError.invalid("WireGuard 설정에 [Peer] 항목이 하나 이상 필요합니다.")
        }

        let config = WireGuardConfig(
            interface: try parseInterface(interfaceValues),
            peers: peers
        )
        try validate(config)
        return config
    }

    nonisolated static func prepareImport(_ configText: String) throws -> WireGuardImportResult {
        let config = try parse(configText)

        return WireGuardImportResult(
            config: config,
            sanitizedConfiguration: sanitize(configText),
            privateKey: config.interface.privateKey,
            summary: createSummary(config)
        )
    }

    nonisolated static func sanitize(_ configText: String) -> String {
        guard !configText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return ""
        }

        return normalizeLineEndings(configText)
            .split(separator: "\n", omittingEmptySubsequences: false)
            .map { rawLine in
                let line = String(rawLine).trimmingCharacters(in: CharacterSet(charactersIn: "\r"))
                let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)

                if trimmed.hasCaseInsensitivePrefix("PrivateKey"), let split = splitKeyValue(trimmed) {
                    return "\(split.key) = <stored securely>"
                }

                return line
            }
            .joined(separator: "\n")
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    nonisolated static func createSummary(_ config: WireGuardConfig) -> WireGuardConfigSummary {
        WireGuardConfigSummary(
            hasInterfaceSection: true,
            hasPeerSection: !config.peers.isEmpty,
            hasPrivateKey: !config.interface.privateKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
            interfaceAddresses: config.interface.addresses,
            dnsServers: config.interface.dnsServers,
            peerCount: config.peers.count,
            allowedIPs: Array(Set(config.peers.flatMap(\.allowedIPs))).sorted(),
            endpoints: config.peers.compactMap(\.endpoint)
        )
    }

    private nonisolated static func parseInterface(_ section: ParsedSection) throws -> WireGuardInterfaceConfig {
        let values = section.values

        return WireGuardInterfaceConfig(
            privateKey: try requiredValue(values, key: "PrivateKey", sectionName: interfaceSection),
            addresses: splitCSV(values["address"]),
            dnsServers: splitCSV(values["dns"]),
            listenPort: try parseNullableInt(values["listenport"], key: "ListenPort")
        )
    }

    private nonisolated static func parsePeer(_ section: ParsedSection) throws -> WireGuardPeerConfig {
        let values = section.values

        return WireGuardPeerConfig(
            publicKey: try requiredValue(values, key: "PublicKey", sectionName: peerSection),
            presharedKey: values["presharedkey"],
            endpoint: values["endpoint"],
            allowedIPs: splitCSV(values["allowedips"]),
            persistentKeepalive: try parseNullableInt(values["persistentkeepalive"], key: "PersistentKeepalive")
        )
    }

    private nonisolated static func validate(_ config: WireGuardConfig) throws {
        guard !config.interface.addresses.isEmpty else {
            throw WireGuardConfigError.invalid("[Interface]에 Address가 하나 이상 필요합니다.")
        }

        try validateBase64LikeKey(config.interface.privateKey, key: "PrivateKey")

        for address in config.interface.addresses {
            try validateCIDR(address, key: "Address")
        }

        for dnsServer in config.interface.dnsServers where IPv4Address(dnsServer) == nil && IPv6Address(dnsServer) == nil {
            throw WireGuardConfigError.invalid("DNS 값이 올바른 IP 주소가 아닙니다: \(dnsServer)")
        }

        for peer in config.peers {
            try validateBase64LikeKey(peer.publicKey, key: "PublicKey")

            if let presharedKey = peer.presharedKey, !presharedKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                try validateBase64LikeKey(presharedKey, key: "PresharedKey")
            }

            guard !peer.allowedIPs.isEmpty else {
                throw WireGuardConfigError.invalid("[Peer]에 AllowedIPs 값이 하나 이상 필요합니다.")
            }

            for allowedIP in peer.allowedIPs {
                try validateCIDR(allowedIP, key: "AllowedIPs")
            }
        }
    }

    private nonisolated static func parseSections(_ configText: String) throws -> [ParsedSection] {
        var sections: [ParsedSection] = []
        var currentSection: ParsedSection?

        for (offset, rawLine) in normalizeLineEndings(configText).split(separator: "\n", omittingEmptySubsequences: false).enumerated() {
            let lineNumber = offset + 1
            let line = stripComment(String(rawLine)).trimmingCharacters(in: .whitespacesAndNewlines)

            guard !line.isEmpty else {
                continue
            }

            if line.hasPrefix("["), line.hasSuffix("]") {
                let sectionName = String(line.dropFirst().dropLast()).trimmingCharacters(in: .whitespacesAndNewlines)
                guard !sectionName.isEmpty else {
                    throw WireGuardConfigError.invalid("\(lineNumber)번째 줄의 섹션 이름이 비어 있습니다.")
                }

                currentSection = ParsedSection(name: sectionName)
                sections.append(currentSection!)
                continue
            }

            guard currentSection != nil else {
                throw WireGuardConfigError.invalid("\(lineNumber)번째 줄의 키/값이 섹션보다 먼저 나왔습니다.")
            }

            guard let split = splitKeyValue(line) else {
                throw WireGuardConfigError.invalid("\(lineNumber)번째 줄의 키/값 형식이 올바르지 않습니다.")
            }

            currentSection?.values[split.key.lowercased()] = split.value
            sections[sections.count - 1] = currentSection!
        }

        return sections
    }

    private nonisolated static func requiredSection(_ sections: [ParsedSection], named name: String) throws -> ParsedSection {
        guard let section = sections.first(where: { $0.name.caseInsensitiveCompare(name) == .orderedSame }) else {
            throw WireGuardConfigError.invalid("WireGuard 설정에 [\(name)] 섹션이 필요합니다.")
        }

        return section
    }

    private nonisolated static func requiredValue(_ values: [String: String], key: String, sectionName: String) throws -> String {
        guard let value = values[key.lowercased()], !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw WireGuardConfigError.invalid("[\(sectionName)]에 \(key) 값이 필요합니다.")
        }

        return value
    }

    private nonisolated static func splitKeyValue(_ line: String) -> (key: String, value: String)? {
        guard let equalsIndex = line.firstIndex(of: "="), equalsIndex != line.startIndex else {
            return nil
        }

        let key = line[..<equalsIndex].trimmingCharacters(in: .whitespacesAndNewlines)
        let value = line[line.index(after: equalsIndex)...].trimmingCharacters(in: .whitespacesAndNewlines)
        return key.isEmpty ? nil : (key, value)
    }

    private nonisolated static func splitCSV(_ value: String?) -> [String] {
        guard let value, !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return []
        }

        return value
            .split(separator: ",")
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
    }

    private nonisolated static func parseNullableInt(_ value: String?, key: String) throws -> Int? {
        guard let value, !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return nil
        }

        guard let parsed = Int(value), parsed >= 0, parsed <= 65_535 else {
            throw WireGuardConfigError.invalid("\(key) 값은 올바른 포트 또는 간격 정수여야 합니다.")
        }

        return parsed
    }

    private nonisolated static func validateBase64LikeKey(_ value: String, key: String) throws {
        guard value.count >= 32 else {
            throw WireGuardConfigError.invalid("\(key) 값이 올바른 WireGuard 키로 사용하기에 너무 짧습니다.")
        }
    }

    private nonisolated static func validateCIDR(_ value: String, key: String) throws {
        let parts = value.split(separator: "/", omittingEmptySubsequences: false)
        guard parts.count == 2, let prefix = Int(parts[1]) else {
            throw WireGuardConfigError.invalid("\(key) 값은 CIDR 형식이어야 합니다: \(value)")
        }

        let address = String(parts[0])
        let maxPrefix: Int
        if IPv4Address(address) != nil {
            maxPrefix = 32
        } else if IPv6Address(address) != nil {
            maxPrefix = 128
        } else {
            throw WireGuardConfigError.invalid("\(key) 값은 CIDR 형식이어야 합니다: \(value)")
        }

        guard prefix >= 0, prefix <= maxPrefix else {
            throw WireGuardConfigError.invalid("\(key) 접두사 범위를 벗어났습니다: \(value)")
        }
    }

    private nonisolated static func stripComment(_ line: String) -> String {
        guard let hashIndex = line.firstIndex(of: "#") else {
            return line
        }

        return String(line[..<hashIndex])
    }

    private nonisolated static func normalizeLineEndings(_ value: String) -> String {
        value.replacingOccurrences(of: "\r\n", with: "\n").replacingOccurrences(of: "\r", with: "\n")
    }
}

private struct ParsedSection {
    let name: String
    var values: [String: String] = [:]
}

private extension String {
    nonisolated func hasCaseInsensitivePrefix(_ prefix: String) -> Bool {
        range(of: prefix, options: [.caseInsensitive, .anchored]) != nil
    }
}
